From b765acf84d3a72530a363270585f5c14f0090173 Mon Sep 17 00:00:00 2001
From: David Ahern <dsahern@gmail.com>
Date: Fri, 7 Jun 2019 15:38:09 -0700
Subject: [PATCH 3/4] libnetlink: Add helper to add a group via setsockopt

groups > 31 have to be joined using the setsockopt. Since the nexthop
group is 32, add a helper to allow 'ip monitor' to listen for nexthop
messages.

Signed-off-by: David Ahern <dsahern@gmail.com>
---
 libnetlink.c | 6 ++++++
 libnetlink.h | 2 ++
 2 files changed, 8 insertions(+)

diff --git a/libnetlink.c b/libnetlink.c
index 97b4f20..d6e21d4 100644
--- a/libnetlink.c
+++ b/libnetlink.c
@@ -37,6 +37,12 @@
 #define RTNL_RCV_BUFSIZE	DEFAULT_RTNL_BUFSIZE
 #endif
 
+int rtnl_add_nl_group(struct rtnl_handle *rth, unsigned int group)
+{
+	return setsockopt(rth->fd, SOL_NETLINK, NETLINK_ADD_MEMBERSHIP,
+			  &group, sizeof(group));
+}
+
 void rtnl_close(struct rtnl_handle *rth)
 {
 	close(rth->fd);
diff --git a/libnetlink.h b/libnetlink.h
index 03ba43e..3de7215 100644
--- a/libnetlink.h
+++ b/libnetlink.h
@@ -17,6 +17,8 @@ struct rtnl_handle
 int rtnl_open(struct rtnl_handle *rth, unsigned subscriptions);
 int rtnl_open_byproto(struct rtnl_handle *rth, unsigned subscriptions,
                       int protocol);
+int rtnl_add_nl_group(struct rtnl_handle *rth, unsigned int group)
+	__attribute__((warn_unused_result));
 void rtnl_close(struct rtnl_handle *rth);
 int rtnl_wilddump_request(struct rtnl_handle *rth, int fam, int type);
 int rtnl_dump_request(struct rtnl_handle *rth, int type, void *req, int len);
-- 
2.30.0

